### PR TITLE
[App Config] `az appconfig feature`: Improve error handling for invalid feature flags

### PR DESCRIPTION
**Pull request by albertofori**
**Related command**

- `az appconfig feature list`
- `az appconfig feature show`

**Description**

- This change updates the `list` command to only filter out invalid feature flags and not error out when an invalid feature flag is found in the list. 
- Updates error handling for `show` command when an invalid feature flag is queried. 

**Testing Guide**

**History Notes**

[App Config] `az appconfig feature`: Improve error handling for invalid feature flags

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [X] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [X] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [X] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
App Config

**Comment by albertofori** [comment]
@avanigupta Could you please take a look at this when you have a chance? Thanks!

File: src/azure-cli/azure/cli/command_modules/appconfig/feature.py [hunk 1]
```diff
@@ -264,6 +264,8 @@ def show_feature(cmd,
         raise CLIErrors.ResourceNotFoundError("Feature '{}' with label '{}' does not exist.".format(feature, label))
     except HttpResponseError as exception:
         raise CLIErrors.AzureResponseError(str(exception))
+    except Exception as exception:
+        raise CLIError(str(exception))
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/feature.py:268]: Why do we catch base Exception and rethrow as CLIError ?

File: src/azure-cli/azure/cli/command_modules/appconfig/_featuremodels.py [hunk 2]
```diff
@@ -319,8 +319,9 @@ def map_keyvalue_to_featureflagvalue(keyvalue):
                     "key:{0} value:{1}\nFull exception: \n{2}".format(keyvalue.key, keyvalue.value, str(exception))
         raise ValueError(error_msg)
 
-    except:
-        logger.error("Exception while parsing feature flag. key:%s value:%s.", keyvalue.key, keyvalue.value)
-        raise
+    except Exception as exception:
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/_featuremodels.py:322]: Its generally bad practice to catch and rethrow base exception by wrapping it into another exception. We should just log the exception and rethrow without wrapping it into ValueError

File: src/azure-cli/azure/cli/command_modules/appconfig/feature.py [hunk 3]
```diff
@@ -293,10 +295,21 @@ def list_feature(cmd,
...
+        invalid_ffs = 0
         for kv in retrieved_keyvalues:
-            retrieved_featureflags.append(
-                map_keyvalue_to_featureflag(
-                    keyvalue=kv, show_conditions=True))
+            try:
+                retrieved_featureflags.append(
+                    map_keyvalue_to_featureflag(
+                        keyvalue=kv, show_conditions=True))
+            except Exception as exception:
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/feature.py:305]: We should catch ValueError only.

File: src/azure-cli/azure/cli/command_modules/appconfig/_featuremodels.py [hunk 4]
```diff
@@ -319,6 +320,11 @@ def map_keyvalue_to_featureflagvalue(keyvalue):
                     "key:{0} value:{1}\nFull exception: \n{2}".format(keyvalue.key, keyvalue.value, str(exception))
         raise ValueError(error_msg)
 
+    except (InvalidArgumentValueError, TypeError) as exception:
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/_featuremodels.py:323]: Can we catch these in the same block as ValueError above? The cause is same for all cases and it should be okay to throw ValueError for all 3 error types.

File: src/azure-cli/azure/cli/command_modules/appconfig/feature.py [hunk 5]
```diff
@@ -3,7 +3,7 @@
 # Licensed under the MIT License. See License.txt in the project root for license information.
 # --------------------------------------------------------------------------------------------
 
-# pylint: disable=line-too-long, too-many-statements, too-many-locals, too-many-branches
+# pylint: disable=line-too-long, too-many-statements, too-many-locals, too-many-branches, broad-except
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/feature.py:6]: Is `broad-except` still needed?

File: src/azure-cli/azure/cli/command_modules/appconfig/feature.py [hunk 6]
```diff
@@ -293,10 +295,21 @@ def list_feature(cmd,
...
         for kv in retrieved_keyvalues:
-            retrieved_featureflags.append(
-                map_keyvalue_to_featureflag(
-                    keyvalue=kv, show_conditions=True))
+            try:
+                retrieved_featureflags.append(
+                    map_keyvalue_to_featureflag(
+                        keyvalue=kv, show_conditions=True))
+            except (CLIErrors.InvalidArgumentValueError, ValueError, TypeError) as exception:
+                logger.warning("%s\n", exception)
```
Comment (avanigupta) [inline src/azure-cli/azure/cli/command_modules/appconfig/feature.py:306]: The error is already being logged in map_keyvalue_to_featureflag. Do we need to log here again?

File: src/azure-cli/azure/cli/command_modules/appconfig/feature.py (hunk 1 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/feature.py:268]: Discussed offline, this is to ensure that the full stack trace of the error is not logged and only the direct cause is shown.

File: src/azure-cli/azure/cli/command_modules/appconfig/_featuremodels.py (hunk 2 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/_featuremodels.py]: This has been updated to only catch specific exceptions, thanks!

File: src/azure-cli/azure/cli/command_modules/appconfig/feature.py (hunk 3 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/feature.py]: This has been updated to catch only the `ValueError` only, thanks!

File: src/azure-cli/azure/cli/command_modules/appconfig/_featuremodels.py (hunk 4 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/_featuremodels.py]: I think this is a good idea. This has been updated accordingly, thanks! Feature.py will catch and log `ValueError` only

File: src/azure-cli/azure/cli/command_modules/appconfig/feature.py (hunk 5 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/feature.py]: My bad. This is not needed anymore. Thanks for pointing this out.

File: src/azure-cli/azure/cli/command_modules/appconfig/feature.py (hunk 6 above)
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/feature.py:306]: Only the unknown exceptions which are not caught here are logged in `map_keyvalue_to_featureflagvalue`. Could I be missing something?

File: src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py [hunk 7]
```diff
@@ -583,11 +583,23 @@ def __serialize_kv_list_to_comparable_json_object(keyvalues, level):
...
+
+        except ValueError:
+            invalid_ffs += 1
+
+    if invalid_ffs > 0:
+        if is_dest:
+            logger.warning("\n%i invalid feature flags from the source configuration could not be previewed but will be imported.", invalid_ffs)
+
+        else:
+            logger.warning("%i invalid feature flags from the destination configuration could not be previewed.", invalid_ffs)
```
[the remark this answers is not in the document]
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/_kv_helpers.py:602]: My bad! I swapped the error messages. Thanks for catching this. This has been updated. Thanks!

**Review by avanigupta — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**